### Heroes Of Pymoli Data Analysis
 

In [53]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

purchase_data.head(10)
#purchase_data.dtypes

Purchase ID             SN  Age                 Gender  Item ID  \
0            0        Lisim78   20                   Male      108   
1            1    Lisovynya38   40                   Male      143   
2            2     Ithergue48   24                   Male       92   
3            3  Chamassasya86   24                   Male      100   
4            4      Iskosia90   23                   Male      131   
5            5        Yalae81   22                   Male       81   
6            6      Itheria73   36                   Male      169   
7            7    Iskjaskst81   20                   Male      162   
8            8      Undjask33   22                   Male       21   
9            9    Chanosian48   35  Other / Non-Disclosed      136   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44  
5                                  Dreamkiss   3.61  
6     Interrogator, Blood Blade of the Queen   2.18  
7                              Abyssal Shard   2.67  
8                                  Souleater   1.10  
9               Ghastly Adamantite Protector   3.58

## Player Count

* Display the total number of players


In [11]:
count_players = len(purchase_data["SN"].unique())
summary_table = pd.DataFrame({"Total players": [count_players]})
summary_table

Total players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [12]:
uniq_items = len(purchase_data["Item Name"].unique())
avg_price = ('${:,.2f}'.format(purchase_data["Price"].mean()))
num_pur = (purchase_data["Purchase ID"].count())
tot_rev = ('${:,.2f}'.format(purchase_data["Price"].sum()))
 
#num_pur, tot_rev
#purchase_data["avg_price"] = purchase_data["avg_price"].astype(float).map("${:,.2f}".format)
#purchase_data["avg_price"] = purchase_data["avg_price"].map("${:,.2f}".format)
#file_df["avg_cost"] = file_df["avg_cost"].map("${:.2f}".format)
#file_df["population"] = file_df["population"].map("{:,}".format)              

summary_table2 = pd.DataFrame({"Number of Unique Items": [uniq_items], "Average Price":[avg_price], "Number of Purchases":[num_pur],"Total Revenue":[tot_rev]})
summary_table2

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [13]:
#gender_groups = purchase_data.groupby("Gender").value_counts()
gender_groups = purchase_data["Gender"].value_counts()
total = purchase_data["Gender"].count()
percent = ((gender_groups /total)*100)
#format_percent = ("{0:.2f}%".format(purchase_data["percent"]))


#percent
#num_pur = purchase_data["Purchase ID"].count()
#num_pur = purchase_data["Purchase ID"].count()
#gendepercent = ([gender_groups] / [num_pur]())
#gender_table = ({"Total Count":[percent]})
gender_table = pd.DataFrame({"Total Count":(gender_groups), "Total %":(percent)})
gender_table
#gender_groups 

Total Count    Total %
Male                           652  83.589744
Female                         113  14.487179
Other / Non-Disclosed           15   1.923077


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
#purchase analysis frame
#create group by gender
grouped_gender = purchase_data.groupby(['Gender'])
#count number of purchases
purchase_count = grouped_gender["Purchase ID"].count()
#putting price into gender grouping
price_sum = grouped_gender["Price"].sum()
#getting average of price by group
price_avg = purchase_data['Price'].mean()
#calculating average price by group
avg_price = (price_sum / purchase_count)

gender_totals = purchase_data["Gender"].value_counts()
Average_Purchase_Total =  price_sum/gender_totals


#create frame
purch_table = pd.DataFrame({"Purchase Count":purchase_count, "Average Purchase Price":avg_price, 
                            "Total Purchase Value":price_sum, "Avg Total Purchase per Person":Average_Purchase_Total})
#assign formatting
purch_table["Average Purchase Price"] = purch_table["Average Purchase Price"].map("${:.2f}".format) 
purch_table["Total Purchase Value"] = purch_table["Total Purchase Value"].map("${:.2f}".format)
purch_table["Avg Total Purchase per Person"] = purch_table["Avg Total Purchase per Person"].map("${:.2f}".format)
purch_table

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $3.20  
Male                              $1967.64                         $3.02  
Other / Non-Disclosed               $50.19                         $3.35

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [14]:
#create the bin
age_bin = [0, 9, 14, 19, 24, 29, 34, 39, 130]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39","40+"]
purchase_data["Age_Ranges"] = pd.cut(purchase_data["Age"], age_bin, labels = group_names)
#calculate unique players - give a total
total_players = len(purchase_data["SN"].unique())
#calculate unique players using drop - removes from the list
purch_without_duplicates = purchase_data.drop_duplicates(["SN"]) 
#create group from bin
ttl_age = purch_without_duplicates.groupby("Age_Ranges")
#put counts into group
total_age = (ttl_age["SN"].count())
#put percents into group and format
pct_age = ((ttl_age["SN"].count()/total_players)).map("{:.2%}".format)

#make the frame
purch_age = pd.DataFrame({"Age Count":total_age, "Age Percent":pct_age})

purch_age                       


Age Count Age Percent
Age_Ranges                       
<10                17       2.95%
10-14              22       3.82%
15-19             107      18.58%
20-24             258      44.79%
25-29              77      13.37%
30-34              52       9.03%
35-39              31       5.38%
40+                12       2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [15]:
#EDIT HERE
#Use Bin from the purchase_data data frame by age from above
purchase_data_age= purchase_data.groupby("Age_Ranges")
#count purchases
purc_count_age = purchase_data_age["Item ID"].count()
#determine average price per person (sum all prices / purchase count)
avg_purc_pr_age = (purchase_data_age["Price"].sum()/purc_count_age).map("${:.2f}".format)
#total price broken by age ranges
ttl_purc_price_age = (purchase_data_age["Price"].sum()).map("${:.2f}".format)
#average by age range = total price divided by count f age 
avg_purc_ttl_age = (purchase_data_age["Price"].sum()/total_age).map("${:.2f}".format)
# make the frame
summary_purchase_age = pd.DataFrame({"Purchase Count": purc_count_age, "Average Purchase Price": avg_purc_pr_age,
                                       "Total Purchase Value": ttl_purc_price_age, "Avg Total Purchase per Person": avg_purc_ttl_age})
summary_purchase_age

Purchase Count Average Purchase Price Total Purchase Value  \
Age_Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age_Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [34]:
#Create a summary data frame to hold the results Purchase ID
purchase_data_age= purchase_data.groupby("SN")

purc_count_item = purchase_data_age["Item ID"].count()

#determine average price per person (sum all prices / purchase count)
avg_purc_pr_age = (purchase_data_age["Price"].sum()/purc_count_item).map("${:.2f}".format)
total_purch_val = purchase_data_age["Price"].sum()

#sort in decending order

#put in frame and format
purch_table3 = pd.DataFrame({"Purchase Count":purc_count_item, "Average Purchase Price":avg_purc_pr_age, 
                            "Total Purchase Value":total_purch_val}).sort_values(by='Total Purchase Value', ascending=False)

#assign formatting
purch_table3["Total Purchase Value"] = purch_table3["Total Purchase Value"].map("${:.2f}".format) 


purch_table3.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [42]:
pop_items = pd.DataFrame({'Item ID' : purchase_data['Item ID'], 'Item Name' : purchase_data['Item Name'], 
                             'Item Price' : purchase_data['Price']})
pop_items_group = pop_items.groupby(['Item ID', 'Item Name'])

pop_items_count = pop_items_group.count()['Item Price']
pop_items_count

pop_tot_value = pop_items_group.sum()['Item Price']
pop_tot_value

#Put calculations into a summary dataframe
pop_summary_df = pd.DataFrame({'Purchase Count' : pop_items_count, 'Total Purchase Value': pop_tot_value})
#pop_summary_df.sort_values(by=['Purchase Count'], ascending=False)

pop_merge_df = pd.merge(pop_items_df, pop_summary_df, on=['Item ID'])
pop_merge_df = pop_merge_df.drop_duplicates()
pop_merge_df = pop_merge_df[['Item ID', 'Item Name', 'Purchase Count', 'Item Price', 'Total Purchase Value']]
most_popular_items = pop_merge_df.sort_values(by=['Purchase Count'], ascending=False).head()

#reset the index so that we can format total purchase value and keep data sorted with the new index
new_pop_index = most_popular_items.reset_index(drop=True)
new_pop_index['Total Purchase Value'] = new_pop_index['Total Purchase Value'].astype(float).map("${:,.2f}".format)
new_pop_index.head()

Item ID                                     Item Name  Purchase Count  \
0      178  Oathbreaker, Last Hope of the Breaking Storm              12   
1      108     Extraction, Quickblade Of Trembling Hands               9   
2       82                                       Nirvana               9   
3      145                          Fiery Glass Crusader               9   
4       19                 Pursuit, Cudgel of Necromancy               8   

   Item Price Total Purchase Value  
0        4.23               $50.76  
1        3.53               $31.77  
2        4.90               $44.10  
3        4.58               $41.22  
4        1.02                $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [51]:
most_profitable = pop_merge_df.sort_values(by=['Total Purchase Value'], ascending=False)
most_profitable.head()

Item ID                                     Item Name  Purchase Count  \
120      178  Oathbreaker, Last Hope of the Breaking Storm              12   
87        82                                       Nirvana               9   
469      145                          Fiery Glass Crusader               9   
15        92                                  Final Critic               8   
547      103                                Singed Scalpel               8   

     Item Price  Total Purchase Value  
120        4.23                 50.76  
87         4.90                 44.10  
469        4.58                 41.22  
15         4.88                 39.04  
547        4.35                 34.80

### Heroes Of Pymoli Data Analysis - Mike Magliaro
* Of the 576 active players, the vast majority are male (84% or 652) with 45% of purchases in the age range of 20-24 with a total of $1,114 purchase value of the total $2,380. 
* The highest purchase by one individual was 5 with a total purchase price of $18.96
* The highest purchase game was  "Oathbreaker, Last Hope of the Breaking Storm" with 12 purchases with a $50.77 value



-----